src/QuadForm/Quad/ZLattices.jl

```diff
@@ -918,12 +918,214 @@ end
+    valv = iszero(sp) ? inf : valuation(sp, p)
+    v = v * VZ
+    sp = (v * G * transpose(v))[1,1]
+    valv = iszero(sp) ? inf : valuation(sp, p)
+    @assert valv >= 2
+    v = QQ(1, p) * change_base_ring(QQ,v)
+  else
+    p = ZZ(p)
+    R8 = ResidueRing(ZZ, ZZ(8))
+    R4 = ResidueRing(ZZ, ZZ(4))
+    findzero_mod4 = function(HR)
+      z = R4(0)
+      i = findfirst(==(z), R4.(diagonal(HR)))
+      v = zero_matrix(ZZ, 1, r)
+      if !(i isa Nothing)
+        v[1, i] = 1
+        return true, v
+      else
+        return false, v
+      end
+    end
+    HR8 = change_base_ring(R8, H)
+    ok, v = findzero_mod4(HR8)
+    B = identity_matrix(R8, nrows(H))
+    if !ok
+      D, B = _jordan_2_adic(HR8)
+      ok, v = findzero_mod4(D)
+    end
+    if !ok
+      D, B1 = Hecke._normalize(D, p)
+      B = B1 * B
+      ok, v = findzero_mod4(D)
+    end
+    if !ok
+      D, B1 = _two_adic_normal_forms(D, p, partial = true)
+      B = B1 * B
+      ok, v = _is_isotropic_with_vector_mod4(D)
+      if !ok
+        return true, L
+      end
+    end
+    v = v * B
+    v = map_entries(ZZ, v)
+    v = v * VZ
+    v = QQ(1,2) * change_base_ring(QQ, v)
+  end
+  v = v * basis_matrix(L)
+  B = vcat(basis_matrix(L), v)
+  LL = lattice(ambient_space(L), B, isbasis=false)
+  @assert det(L) ==  det(LL) * p^2 && valuation(norm(LL), p) >= 0
+  @assert denominator(scale(LL))==1
+  @assert mod(ZZ(norm(LL)),2)==0
+  return false, LL
+end
+
+@doc Markdown.doc"""
+    _is_isotropic_with_vector_mod4(Gnormal) -> Bool, MatElem
+
+Return if `Gnormal` is isotropic mod 4 and an isotropic vector.
+
+Assumes that G is in partial 2-adic normal form.
+"""
+function _is_isotropic_with_vector_mod4(Gnormal)
+  R4 = ResidueRing(ZZ, 4)
+  G = change_base_ring(R4, Gnormal)
+  D = diagonal(G)
+  z = R4(0)
+  v = zero_matrix(ZZ, 1, ncols(G))
+  i = findfirst(==(z), D)
+  if !(i isa Nothing)
+    v[1, i] = 1
+    return true, v
+  end
+  @assert nrows(G) <= 6 "$G"
+  if nrows(G) == 1
+    return false, v
+  end
+  # hardcoded isotropic vector for G in normal form (and no 0 mod 4 on the diag)
+  if nrows(G) == 2
+    if G[1,2] == 0 && D[1]+D[2] == 0
+      v[1,1] = 1
+      v[1,2] = 1
+      return true, v
+    else
+      return false, v
+    end
+  end
+  if nrows(G) == 3
+    if D[3] in [R4(1),R4(3)]
+      return false, v
+    end
+    @assert D[3] == R4(2)
+    if sum(D[2:3]) == 0
+      v[1,2] = 1; v[1,3] = 1
+      return true, v
+    end
+    if G[1,2] == 0 && sum(D[1:2]) == 0
+      v[1,1] = 1; v[1,2] = 1
+      return true, v
+    end
+    if G[1,2] == 0 && sum(D) == 0
+      v[1,1] = 1; v[1,2] = 1; v[1,3] = 1
+      return true, v
+    end
+  end
+  n = nrows(G)
+  if D[1]+D[n] == 0
+    v[1,1] = 1
+    v[1,n] = 1
+    return true, v
+  end
+  if D[n-1]+D[n] == 0
+    v[1,n-1] = 1
+    v[1,n] = 1
+    return true, v
+  end
+  if D[1]+D[n-1] + D[n] == 0
+    v[1,1] = 1
+    v[1,n-1] = 1
+    v[1,n] = 1
+    return true, v
+  end
+  @assert false
```

Comment:
```suggestion
  error("Something wrong!")
```
Julia might decide to remove asserts at some later point.